dimod/binary_quadratic_model.py

```diff
@@ -1953,6 +1955,134 @@ def from_numpy_matrix(cls, mat, variable_order=None, offset=0.0, interactions=No
+
+        """
+        linear = self.linear
+        quadratic = self.quadratic
+        n = len(linear)
+
+        if variable_order is None:
+            if not all(v in self.linear for v in range(n)):
+                raise ValueError(("if 'variable_order' is not provided, binary quadratic model must be "
+                                  "index labeled [0, ..., N-1]"))
+            variable_order = list(range(n))
+
+        labels = {v: idx for idx, v in enumerate(variable_order)}
+
+        lin = np.array([linear[v] for v in variable_order], dtype=dtype)
+
+        if sort_indices:
+            def _iter_quadratic():
+                for (u, v), bias in iteritems(quadratic):
+                    u = labels[u]
+                    v = labels[v]
+
+                    if u > v:
+                        yield (v, u, bias)
+                    else:
+                        yield (u, v, bias)
+
+            heads, tails, values = zip(*sorted(_iter_quadratic()))
+        else:
+            heads, tails, values = zip(*((labels[u], labels[v], bias) for (u, v), bias in iteritems(quadratic)))
+
+        return lin, (np.array(heads, dtype=index_dtype),
+                     np.array(tails, dtype=index_dtype),
+                     np.array(values, dtype=dtype)), self.offset
+
+    @classmethod
+    def from_numpy_vectors(cls, linear, quadratic, offset, vartype, variable_order=None):
+        """Create a binary quadratic model from vectors.
+
+        Args:
+            linear (array_like):
+                A 1D array-like iterable of linear biases.
+
+            quadratic (tuple[array_like, array_like, array_like]):
+                A 3-tuple of 1D array_like vectors of the form (row, col, bias).
+
+            offset (numeric, optional):
+                Constant offset for the binary quadratic model.
+
+            vartype (:class:`.Vartype`/str/set):
+                Variable type for the binary quadratic model. Accepted input values:
+
+                * :class:`.Vartype.SPIN`, ``'SPIN'``, ``{-1, 1}``
+                * :class:`.Vartype.BINARY`, ``'BINARY'``, ``{0, 1}``
+
+            variable_order (iterable, optional):
+                If provided, labels the variables; otherwise, indices are used.
+
+        Returns:
+            :obj:`.BinaryQuadraticModel`
+
+        Examples:
+            >>> linear_vector = np.asarray([-1, 1])
+            >>> quadratic_vectors = (np.asarray([0]), np.asarray([1]), np.asarray([-1.0])))
+            >>> bqm = dimod.BinaryQuadraticModel.from_numpy_vectors(linear_vector, quadratic_vectors, 0.0, dimod.SPIN)
+            >>> bqm.quadratic
+            {(0, 1): -1}
+
+        """
+
+        try:
+            heads, tails, values = quadratic
+        except ValueError:
+            raise ValueError("quadratic should be a 3-tuple")
+
+        if variable_order is None:
+            variable_order = list(range(len(linear)))
+
+        linear = {v: bias for v, bias in zip(variable_order, linear)}
```

Comment:
Is it okay that the keys and values of these dicts will retain their `numpy` types? I believe this may cause unexpected behavior for the user (though it should be rare... perhaps something like submitting index arrays of type `np.uint8` and getting overflow later on could be a thing though), and can cause performance issues with dicts (https://stackoverflow.com/a/24705922).

Reply:
Will fix

Reply:
My bad, I didn't notice the keys are actually Python, so the problems I talked about are not relevant.